packages/rest/src/body-parser.ts

```diff
@@ -0,0 +1,287 @@
+  if (!err.statusCode || err.statusCode >= 500) {
+    err.statusCode = 400;
+  }
+  return err;
+}
+
+/**
+ * Parse the body asynchronously
+ * @param handle The express middleware handler
+ * @param request Http request
+ */
+function parse(
+  handle: BodyParserWithCallback,
+  request: Request,
+): Promise<void> {
+  // A hack to fool TypeScript as we don't need `response`
+  const response = ({} as any) as Response;
+  return new Promise<void>((resolve, reject) => {
+    handle(request, response, err => {
+      if (err) {
+        reject(normalizeParsingError(err));
+        return;
+      }
+      resolve();
+    });
+  });
+}
+
+// Default limit of the body length
+const DEFAULT_LIMIT = '1mb';
+
+type ParserOption<T extends 'json' | 'urlencoded' | 'text'> = T extends 'json'
+  ? OptionsJson
+  : T extends 'urlencoded'
+    ? OptionsUrlencoded
+    : T extends 'text' ? OptionsText : Options;
+
+function getParserOptions<T extends 'json' | 'urlencoded' | 'text'>(
+  type: T,
+  options: RequestBodyParserOptions,
+): ParserOption<T> {
+  const opts: {[name: string]: any} = {};
+  Object.assign(opts, options[type], options);
+  for (const k of ['json', 'urlencoded', 'text']) {
+    delete opts[k];
+  }
+  return opts as ParserOption<T>;
+}
+
+/**
+ * Interface to be implemented by body parser extensions
+ */
+export interface BodyParser {
+  /**
+   * Optional name of the parser for debugging
+   */
+  name?: string;
+  /**
+   * Indicate if the given media type is supported
+   * @param mediaType Media type
+   */
+  supports(mediaType: string): boolean;
+  /**
+   * Parse the request body
+   * @param request http request
+   */
+  parse(request: Request): Promise<RequestBody>;
+}
+
+/**
+ * Binding tag for request body parser extensions
+ */
+export const REQUEST_BODY_PARSER_TAG = 'rest.requestBodyParser';
+
+export class RequestBodyParser {
+  private parsers: BodyParser[];
+  private jsonParser: BodyParserWithCallback;
+  private urlencodedParser: BodyParserWithCallback;
+  private textParser: BodyParserWithCallback;
+
+  constructor(
+    @inject(RestBindings.REQUEST_BODY_PARSER_OPTIONS, {optional: true})
+    options: RequestBodyParserOptions = {},
+    @inject.tag(REQUEST_BODY_PARSER_TAG, {optional: true})
+    parsers: BodyParser[] = [],
+  ) {
+    const jsonOptions = Object.assign(
+      {type: 'json', limit: DEFAULT_LIMIT},
+      getParserOptions('json', options),
+    );
+    this.jsonParser = json(jsonOptions);
+
+    const urlencodedOptions = Object.assign(
+      {
+        type: 'urlencoded',
+        extended: true,
+        limit: DEFAULT_LIMIT,
+      },
+      getParserOptions('urlencoded', options),
+    );
+    this.urlencodedParser = urlencoded(urlencodedOptions);
+
+    const textOptions = Object.assign(
+      {type: 'text/*', limit: DEFAULT_LIMIT},
+      getParserOptions('text', options),
+    );
+    this.textParser = text(textOptions);
+
+    this.parsers = parsers.concat([
```

Comment:
1. Can applications disable certain body parsers? If not, can they provide custom implementation for given media types?

2. I find this implementation a bit hacky in the sense that it has different internal structure from what parsers contributed by extensions would use. Please rework it so that internal parsers look the same way as external parsers would do. Ideally, each built-in parser should have its own source code file, e.g. `packages/rest/src/body-parser/json.body-parser.ts`.

I hate to repeat myself so often, but: it would be much easier for both of us if one pull request contained one logical change only.

Reply:
1 & 2 are fixed.

I'm sorry that I had to combine a few things in a single PR as they build on top of each other. Maintaining 3 PRs (such as pr3 -> pr2 --> pr1) is hard for me too. I intentionally keep them as separate commits.